### ISSUE TITLE
ChainerX grads are always retained when using chainer.

### ISSUE THREAD

**Issue by emcastillo** [chainer/chainer#7587]
When using chainer with chainerx,
If the input variables are flagged with `requires_grad=True`, then all of the intermediate variables in the graph resulting from calling functions on the input values are being created calling the corresponding `require_grad` of chainerx.

The issue here is that in this case, all the chainerx variables will retain the calculated grads no matter what, as opposed to chainer, where the variables do not retain the grad unless it is expliclity specified.

`function_node.py:281`
when doing forward computation of chainerx variables
it creates a variable to wrap the output array with `require_grad` set
```
if outputs is not chainer.Fallback:
     # Supported. Wrap with variables and return
     assert isinstance(outputs, tuple)
     return tuple([
                    variable.Variable._init_unchecked(
                        y, requires_grad=y.is_backprop_required(),
                        is_chainerx_array=True)
                    for y in outputs])
```

`Variable` `__init__` calls _`_set_chainerx_array` which does the following over the chainerx array

```
if requires_grad:
      array.require_grad()
      if grad is not None:
          array.set_grad(grad)
```

require_grad on a chainer array forces it to retain the calculated grad.

**Comment by niboshi** [comment]
In order to investigate this issue can actually causes memory consumption problem, I conducted a test with the following script:

```py
import chainerx
import chainer
from chainer import functions as F
import os, psutil


def get_mem():
    return psutil.Process(os.getpid()).memory_info().rss

base_mem = get_mem()

def print_mem(label):
    print(label, ': ', (get_mem() - base_mem) // 1024 // 1024)

shape = (1024 * 1024 * 25)
dtype = 'float32'
device = chainerx.get_device('native:0')


def func(a):
    #return a * a  # retain input
    return F.exp(a)  # retain output

with chainerx.using_device(device):
    print_mem('A')
    x = chainer.Variable(chainerx.ones(shape, dtype))
    print_mem('B')
    y = func(x)
    print_mem('C')
    z = func(y)
    print_mem('D')
    z.grad = chainerx.ones(shape, dtype)
    print_mem('E')
    z.backward()
    print_mem('F')
    del y
    print_mem('G')

```

`a * a` for the case of input retention, and `F.exp(a)` for output retention.
Results follow:
### Input is retained
```
A :  0
B :  100
C :  200
D :  300
E :  400
F :  600
G :  400
```

###  Output is retained
```
A :  0
B :  100
C :  200
D :  300
E :  400
F :  600
G :  400
```

So in both cases, it's confirmed that
* intermediate grads were indeed retained after backward (consumption at F was increased by two arrays (gy and gx)), and
* the retained grad was freed by `del`ing y (at G). At this point the memory worth two arrays were freed (presumably y and gy).

**Comment by niboshi** [comment]
In order to measure peak memory consumption during backprop in typical use case where intermediate array `y` is gone during forward propagation, I updated the script like below.
I added the no-retain case.

```py
import chainerx
import chainer
from chainer import functions as F


shape = (1024 * 1024 * 25)
dtype = 'float32'
device = chainerx.get_device('cuda:0')


def func(a):
    return 1 + a  # no retain
    #return F.log(a)  # retain input
    #return F.exp(a)  # retain output

with chainerx.using_device(device):

    chainerx.ones((1,), 'bool')  # initialize CUDA context

    input('Measure base memory consumption: ')

    x = chainer.Variable(chainerx.ones(shape, dtype))
    y = x
    for _ in range(10):
        y = func(y)

    input('Measure memory consumption after forward: ')

    y.grad = chainerx.ones(shape, dtype)
    y.backward()

    input('Measure peak memory consumption: ')
```
Thanks to CUDA memory pool, I could measure the peak memory consumption at each step.
(all numbers in MiB)

| | no retain | input retain | output retain |
|---|---|---|---|
| Base | 445 | 445 | 445 |
| After forward | 745 | 1545 | 1545 |
| Peak | 745 | 1745 | 1845 |
| Delta during backprop | 0 | 200 | 300 |

None of the three cases consumed something comparable to 10 * 100 = 1000 MiB.
It is confirmed that intermediate gradients are freed during backprop if intermediate arrays are not kept.

**Comment by hvy** [comment]
The issue is still present about retained grads.
